pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,517 @@
+	if config.frontend == nil {
+		return
+	}
+	found, err := rpm.svcManager.DeleteService(*config.frontend)
+	delete(rpm.frontends, config.frontend.Hash())
+	config.frontend = nil
+	if !found || err != nil {
+		log.WithError(err).Debugf("Local redirect service for policy %s not deleted",
+			config.getPolicyID())
+	}
+}
+
+// Updates datapath with the backend deletion.
+func (rpm *RedirectPolicyManager) notifyPolicyBackendDelete(config *LRPConfig) {
+	if config.frontend.IP != nil {
+		if len(config.backends) > 0 {
+			rpm.upsertService(config)
+		} else {
+			// No backends so remove the service entry.
+			found, err := rpm.svcManager.DeleteService(*config.frontend)
+			if !found || err != nil {
+				log.WithError(err).Debugf("Local redirect service for policy %s not deleted",
+					config.getPolicyID())
+			}
+		}
+	}
+}
+
+// Checks if the given policy config is ready and plumbs datapath with a service
+// entry created from the policy config.
+func (rpm *RedirectPolicyManager) notifyPolicyConfigUpdate(config *LRPConfig) {
+	if config.frontend.IP != nil && len(config.backends) > 0 {
+		// Policy config is ready.
+		rpm.upsertService(config)
+	}
+}
+
+func (rpm *RedirectPolicyManager) upsertService(config *LRPConfig) {
+	frontendAddr := lb.L3n4AddrID{
+		L3n4Addr: *config.frontend,
+		ID:       lb.ID(0),
+	}
+	var backendAddrs []lb.Backend
+	for _, be := range config.backends {
+		backendAddrs = append(backendAddrs, lb.Backend{
+			NodeName: nodeTypes.GetName(),
+			L3n4Addr: be,
+		})
+	}
+	p := &lb.SVC{
+		Name:          config.name + localRedirectSvcStr,
+		Namespace:     config.namespace,
+		Type:          lb.SVCTypeLocalRedirect,
+		Frontend:      frontendAddr,
+		Backends:      backendAddrs,
+		TrafficPolicy: lb.SVCTrafficPolicyCluster,
+	}
+	log.Info("Upserting a service for redirect policy config %s with "+
+		"params %#v", config.getPolicyID(), p)
+	if _, id, err := rpm.svcManager.UpsertService(p); err != nil {
+		log.WithError(err).Error("Error while inserting service in LB map")
+	} else {
+		log.Infof("Service created for local redirect policy %s with ID %d",
+			config.getPolicyID(), id)
+	}
+}
+
+// Returns a slice of endpoint pods metadata that are selected by the given policy config.
+func (rpm *RedirectPolicyManager) getLocalPodsForPolicy(config *LRPConfig, epManager endpointManager) []*epPodInfo {
+	var retEps []*epPodInfo
+
+	for _, ep := range epManager.GetEndpoints() {
+		podMetadata := rpm.getEndpointPodMetadata(ep)
+		if podMetadata == nil {
+			// TODO Pod metadata isn't available yet. We need to add this to
+			// pending request.
+			continue
+		}
+		if !config.policyConfigSelectsPod(podMetadata) {
+			continue
+		}
+		retEps = append(retEps, podMetadata)
+	}
+
+	return retEps
+}
+
+// Returns backend entries for pod(s) for the given redirect policy config.
+func (rpm *RedirectPolicyManager) getPodBackendsForPolicy(config *LRPConfig, podsInfo ...*epPodInfo) []backend {
```

Comment:
It seems like we don't check for backend status? In the case of node-local-dns, if backend is not in an healthy status, enforcing LRP would break DNS?

Reply:
I checked service code that adds backends for services of type ClusterIP, and we don't seem to check backend status before adding it to the map. Regardless, this is something that needs to be done periodically, right? Are you thinking of using Pod readiness/liveness probes to determine its status? We seem to maintain a `HealthCheckNodePort`, but looks like that's only used for external load balancer services. @brb Did I miss anything?

Reply:
Also, how is this handled for a regular node-local dns environment?

Reply:
Based on the Kubernetes guide [1], if pods are deployed with readiness/liveness probes, then kubelet is responsible for killing and re-spawning unhealthy pods so that kube-proxy never sees such backends. Similarly, cilium agent should get pod delete events via watcher, it would then go ahead and remove that backend from the map?

[1] https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service

Reply:
Node-local dns by itself simply listens on dns cluster ip when it's up, so when it is not up, (since there's no packet rewriting) dns requests simply go to kube-dns.

Reply:
In any case, this is out of scope of this PR. We can discuss this further offline/sig-datapath meeting.

Reply:
Per offline discussion [1], Weilong will be sending a PR to account for backend pod's health status before and while it's added to the BPF maps.

[1] https://cilium.slack.com/archives/C1MATJ5U5/p1598378691128400

Reply:
Agree, I think this is out-of-scope for this PR. By all means let's follow up to make sure whether it makes sense to address and how we want to approach it. (I haven't caught up on the latest Slack thread)